### Pandas Lab -- Cleaning, Merging, & Grouping

This lab is designed to introduce students to common use cases for Pandas when working with data:

 - Creating new information out of your existing data set
 - Merging, concatenating, and joining different data sources
 - Grouping -- With both time & non-time based data

### Section I: Creating Data Out of Your Existing Columns

Go ahead and create the following columns in your dataset.

**Column 1:**

  - **Column Name:** Weekend
  - **Values:** `True` if `day_of_week` is either Friday or Saturday, `False` if not

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv('/Users/cdeboer/Desktop/GA/GA Local/ClassMaterial/Unit2/data/restaurant_data/master.csv')

df['weekend'] = np.where((df['day_of_week'] == 'Saturday') | (df['day_of_week'] == 'Sunday'), True, False)

In [4]:
conditions = [df['reserve_visitors'] < df['reserve_visitors'].quantile(0.25),
              df['reserve_visitors'].between(df['reserve_visitors'].quantile(0.25), df['reserve_visitors'].quantile(0.75)),
              df['reserve_visitors'] > df['reserve_visitors'].quantile(0.75)]

results    = [
    'Low',
    'Medium',
    'High'
]

df['Reservation Activity'] = np.select(conditions, results, 'Other')

**Column 2:**

 - **Column Name:** Reservation Activity
 - **Values:**
   - `Low` if `reserve_visitors` is in the bottom .25 percentile
   - `Medium` if `reserve_visitors` is in the middle .50 percentile
   - `High`if `reserve_visitors` is in the top .25 percentile
   
**Hint:** Use the `quantile` method to get this value

In [5]:
df['Reservation Activity']

0          Other
1          Other
2          Other
3          Other
4          Other
           ...  
252103    Medium
252104      High
252105      High
252106       Low
252107      High
Name: Reservation Activity, Length: 252108, dtype: object

**Column 3:**

 - **Column Name:** Days
 - **Values:**
   - The length of time that has passed from the beginning of the time series, in days
 - **Note:** When you subtract these columns, your column will be a **time delta**.  See if you can use the `dt` attribute to convert these values into an integer.  Ie, if your value reads `3 days`, you want that to be 3 instead.  You can read more about different time periods in pandas here:  https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-date-components

In [6]:
air_reserve = pd.read_csv('/Users/cdeboer/Desktop/GA/GA Local/ClassMaterial/Unit2/data/restaurant_data/air_reserve.csv')
air_store_info = pd.read_csv('/Users/cdeboer/Desktop/GA/GA Local/ClassMaterial/Unit2/data/restaurant_data/air_store_info.csv')
air_visit_data = pd.read_csv('/Users/cdeboer/Desktop/GA/GA Local/ClassMaterial/Unit2/data/restaurant_data/air_visit_data.csv')
date_info = pd.read_csv('/Users/cdeboer/Desktop/GA/GA Local/ClassMaterial/Unit2/data/restaurant_data/date_info.csv')

### Section II: Merging Dataframes

The dataset we have been working with so far (`master.csv`), is actually a combined version of several datasets.  

In this section of the lab, we are going to re-create it manually from its individual pieces.

In the `restaurant data` folder, you'll find the following files:

 - `air_reserve.csv`
 - `air_store_info.csv`
 - `air_visit_data.csv`
 - `date_info.csv`
 
They contain all the constituent info for the `master.csv` file that we're currently using. 

You should have 252108 rows when you are finished.

Using merges, piece the files together to recreate the one we are currently working on.  

**Hint:** To get the number of reservations in the `reserve_visitors` column, you will have to use the `groupby` method first for each `store_id` and `day` before doing the merging.

You will also have to make sure each column is the same datatype -- date is probably the best.  Check the amount of null values in the new column to ensure you did it correctly.  (An incorrect merge will have no non-null values).

Some operations that might come in handy:

 - `dt.date` -- converts a datetime to a date
 - `pd.to_datetime` if you need to convert something from a string to a date

In [ ]:
# your answer here